ext/mbstring/libmbfl/filters/mbfilter_utf16.c

```diff
@@ -324,3 +346,108 @@ int mbfl_filt_conv_wchar_utf16le(int c, mbfl_convert_filter *filter)
 
 	return c;
 }
+
+static int mbfl_filt_ident_utf16(int c, mbfl_identify_filter *filter)
+{
+	if (filter->status == 0) {
+		if (c >= 0xfe) { /* could be a byte-order mark */
+			filter->status = c;
+		} else {
+			/* no byte-order mark at beginning of input; assume UTF-16BE */
+			filter->filter_function = mbfl_filt_ident_utf16be;
+			return (filter->filter_function)(c, filter);
+		}
+	} else {
+		unsigned short n = (filter->status << 8) | c;
+		filter->status = 0;
+
+		if (n == 0xfeff) {
+			/* it was a big-endian byte-order mark */
+			filter->filter_function = mbfl_filt_ident_utf16be;
```

Comment:
It appears that the author of libmbfl intended for identify and conversion filters to *both* be reusable; but mbstring never does so. (Note the way that `_flush` functions generally re-initialize fields.) Also, I am planning to convert most uses of identify/conversion filters to use stack allocation rather than heap allocation as at present. After that, the chances will become even smaller that we will ever want to reuse them.

However, if we ever really want to reuse these structures, the best solution would be to have a function which re-initializes the structure correctly. This might be the same as the `ctor` function.

Reply:
Let me also mention that many, many of the libmbfl identify filters have bugs whereby certain strings are wrongly rejected, because `status` is being used to hold state which cannot be cleared every time we reach a legal point to break the string. The simplest way to fix these bugs is to do something similar to what you see here, where part of the state is represented by swapping in a different `filter_function`.

As a bonus, it avoids doing a `switch` to determine the state we are in on each and every call, and this tends to be a tiny bit faster.

Actually, as this code is optimized for higher performance, I expect that both conversion and identification filters will start working on *buffers* of text rather than single characters. At that point, the tradeoffs will become different and it may likely happen that there will be no swapping of `filter_function`s.

The first problem, though, is to get the code *correct*. When I started all this work on mbstring, I imagined that due to its long history, it must be fully debugged already; but I have found that really, it is full of all kinds of bugs. Second problem is to clean up and remove dead and unneeded code. At some point, when the code gets into a better state, it will be time to more aggressively push for performance.

Reply:
Okay, let's go with this for now. We don't need a stable filter function right now, and this does make some things simpler.

I agree that the input filters should ideally be working on a buffer, that should both make faster *and* easier to implement/understand. I'm not sure if we always have a buffer available though (i.e. if there are any places in libmbfl where we receive bytes one by one.)

> When I started all this work on mbstring, I imagined that due to its long history, it must be fully debugged already;

Heh, if only it were so :)